Add tap-tempo configuration and handling for effect presets

a2hsh opened the pull request:
### summary
This pull request introduces tap-tempo support for effect presets, allowing users to set the tempo of echo effects by tapping preset buttons four times, and adds a configurable tap-tempo window. The implementation includes changes to the CLI, device logic, settings management, and IPC commands to support this new feature.

### **Tap-tempo feature implementation:**

* Added a new `TapTempoWindow` variant to the `DeviceSettings` CLI enum, allowing users to configure the tap-tempo window duration (1000–2000 ms) via the CLI.
* Implemented tap-tempo state tracking and logic in `daemon/src/device.rs`, including the `TapTempoState` struct, tap state management per effect preset, and logic to compute BPM from tap intervals. [[1]](diffhunk://#diff-2626e4b8e1385b6a80cda28158e8658fd115e8db2f9b92efbda54927579d2093R63-R66) [[2]](diffhunk://#diff-2626e4b8e1385b6a80cda28158e8658fd115e8db2f9b92efbda54927579d2093R81-R140) [[3]](diffhunk://#diff-2626e4b8e1385b6a80cda28158e8658fd115e8db2f9b92efbda54927579d2093R293-R297) [[4]](diffhunk://#diff-2626e4b8e1385b6a80cda28158e8658fd115e8db2f9b92efbda54927579d2093R824-R877)
* Modified effect preset button handling to support tap-tempo: consecutive taps within the window update the tempo, and effect bank loading suppresses TTS messages on subsequent taps. [[1]](diffhunk://#diff-2626e4b8e1385b6a80cda28158e8658fd115e8db2f9b92efbda54927579d2093L664-R767) [[2]](diffhunk://#diff-2626e4b8e1385b6a80cda28158e8658fd115e8db2f9b92efbda54927579d2093L1475-R1623)

#### **Settings and persistence:**

* Added tap-tempo window configuration to settings: new fields in `DeviceSettings`, default values, and getter/setter methods for device tap-tempo window. [[1]](diffhunk://#diff-ab922ea019ed277aea3988c089f2b431c1e2a925fe80805c643007b5fea95ad3R842-R844) [[2]](diffhunk://#diff-ab922ea019ed277aea3988c089f2b431c1e2a925fe80805c643007b5fea95ad3R867) [[3]](diffhunk://#diff-ab922ea019ed277aea3988c089f2b431c1e2a925fe80805c643007b5fea95ad3R530-R551)
* Updated device and IPC structures to include the tap-tempo window value, ensuring persistence and communication between daemon and client. [[1]](diffhunk://#diff-2626e4b8e1385b6a80cda28158e8658fd115e8db2f9b92efbda54927579d2093R427) [[2]](diffhunk://#diff-fe632566a0e1ef29c3a2a2d4a0526c765faf68d1de7d09a7ee4639854a5bec4cR412)

#### **IPC and command integration:**

* Added a new `SetTapTempoWindow(u16)` command to the IPC layer and integrated its handling in the device daemon, including clamping and saving the window value. [[1]](diffhunk://#diff-66290128b40986b905eff70bf067a8fc8b9f780de7bbde77989e965ad96b4584R311-R312) [[2]](diffhunk://#diff-2626e4b8e1385b6a80cda28158e8658fd115e8db2f9b92efbda54927579d2093R2937-R2945) [[3]](diffhunk://#diff-072974c208ee3614b0a5083397fd515022eb76ef41101521053668cae1835dc3R1025-R1029)

#### **UI Updates:**

* A pull request is created on the [goxlr-ui Repo](https://github.com/GoXLR-on-Linux/goxlr-ui/pull/73) to add the configurable tab tempo window to device settings.

a2hsh committed:
Add tap-tempo configuration and handling for effect presets

FrostyCoolSlug commented:
Thanks for this, it's a feature of the official app I've not really put much thought into, glad to see someone else giving it some time.

I'll give it a proper review in the next couple of days when I can find the time, but just some initial thoughts after a quick browse..

If an effect preset is active, and someone hits the preset button again, there's really no need to reload the bank (this is a small oversight of mine).. So in the button handler it might make more sense to do an 'if active preset != selected, load the preset, else assume a tempo tap', which saves the need to work out whether you have to suppress the default behaviour.

This one is optional, but instead of having a fixed 4 entries in the struct, would a small Vec make sense instead? For each tap you could remove the first entry, and add a new entry to the end, then recalculate, that way the number of taps isn't specifically limited.

Other than that, everything looks great :)

a2hsh committed:
Update TTS message format for BPM delay to be less verbose

a2hsh committed:
Refactor TapTempoState to use a vector for taps and avoid reloading the active preset, avoiding the need to supress TTS Announcements.

a2hsh committed:
Improve error message formatting for ms_window_value function in DeviceSettings

a2hsh committed:
Refactor BPM clamping logic in Device to use clamp method for improved readability

a2hsh commented:
Hi @FrostyCoolSlug 

Thanks for looking into this.
This is officially my first ever contribution in Rust, so thanks for your patience with me ☺️

I have applied your requested changes successfully, CI/CD now also passes:

### Summary                                                                                                                 
                                                                                                                        
- Effect preset buttons: Only reload on preset change; pressing the active preset now counts as a tempo tap.            
- Tap-tempo logic: Refactored from 4 fixed slots to a capped Vec<Instant> (10). Averages all recent intervals within the
window for smoother BPM.                                                                                                
- Continuous refinement: Stops resetting after applying tempo. Recalculates on every subsequent tap until a gap > window
occurs (then starts a fresh sequence).                                                                                  
- Simplicity: Removed the old “suppress default” path and helper. Selection vs tap handled directly in button handlers. 
                                                                                                                        
### Details                                                                                                                 
                                                                                                                        
- File: goxlr-utility/daemon/src/device.rs                                                                              
- Key areas: on_button_up (EffectSelect1..6), TapTempoState, try_tap_tempo                                              
- Build: Verified goxlr-daemon builds cleanly.                                                                          
 
I also fixed a formatting error that was preventing ci/cd from passing earlier.

FrostyCoolSlug commented on client/src/cli.rs:
```diff
@@ -1181,4 +1181,22 @@ pub enum DeviceSettings {
     },
+
+    /// Configure the tap-tempo window (ms) for effect preset tapping
+    TapTempoWindow {
```
There's an obvious question here, is this needed? The maximum possible value for the echo tempo is 300ms, so it makes sense to assume that if they're tapping slower than 300ms then we should just ignore or reset the counter.

FrostyCoolSlug commented on daemon/src/device.rs:
```diff
@@ -60,6 +60,10 @@ pub struct Device<'a> {
+
+    // Tap-tempo configuration and state
+    tap_tempo_window: Duration,
+    tap_states: HashMap<EffectBankPresets, TapTempoState>,
```
Why is this HashMapped specifically to the preset? It's a 'general purpose' handler that doesn't have a relation to the button being pressed, we can simply have one TapTempoState and reset it when the preset changes.

FrostyCoolSlug commented on daemon/src/device.rs:
```diff
@@ -74,6 +78,57 @@ struct ButtonState {
+        }
+
+        // Add the new tap
+        self.taps.push(now);
```
This seems a little weird, would it not be better to create the taps vec with `Vec::with_capacity(4)`, then

```rust
if self.taps.length() == self.taps.capacity() {
  self.taps.remove(0);
}
taps.push(now)
```

This would allow a continuous number of taps without limit, or overflow checking, and you can work on the data available.

FrostyCoolSlug commented on daemon/src/device.rs:
```diff
@@ -74,6 +78,57 @@ struct ButtonState {
+        }
+
+        // Compute durations using all recent intervals once we have at least 4 taps
+        if self.taps.len() >= 4 {
```
Why do we need a minimum of 4 taps? It should be possible to work out the tempo with a minimum of 2 taps on the button, extra taps are useful for refining the granularity.

FrostyCoolSlug commented on daemon/src/device.rs:
```diff
@@ -1472,11 +1611,13 @@ impl<'a> Device<'a> {
-        let preset_name = self.profile.get_effect_name(preset);
-        let tts_message = format!("Effects {}, {}", preset as u8 + 1, preset_name);
-        let _ = self.global_events.send(TTSMessage(tts_message)).await;
+    async fn load_effect_bank(&mut self, preset: EffectBankPresets, speak_tts: bool) -> Result<()> {
```
As previously mentioned, speak_tts isn't needed anymore, It's currently only ever sent as true

FrostyCoolSlug commented on daemon/src/device.rs:
```diff
@@ -2790,6 +2931,15 @@ impl<'a> Device<'a> {
             }
+            GoXLRCommand::SetTapTempoWindow(duration_ms) => {
+                // Clamp to a sensible range (1000ms..=2000ms)
+                let clamped: u16 = duration_ms.clamp(1000, 2000);
```
Assuming this is kept, I'm not a fan of clamping invalid values, if the value is wrong, it should be rejected and not transparently changed under the hood. It's simply not clear to users when they change something and it's not set to the value they sent.

FrostyCoolSlug reviewed (commented):
A few comments here on things which could be improved.

a2hsh commented:
> As an extra note, it seems relatively clear that some "AI" has been used as part of this PR, There's some pretty clear evidence of it in the comments, and while I could be wrong, I'm seeing potential evidence of it in the code.
> 
> Could you please disclose your usage of any LLM used in the generation of this PR?

Hi @FrostyCoolSlug 
Thanks for the detailed review.

As my experience in rust is relatively new, I used AI in writing some functions and to get review feedback, codex-cli to be exact.

For the other comments, I will try to work on them in the next month as time permits.
Thanks!
